cpp/src/text/subword/detail/data_normalizer.hpp

```diff
@@ -0,0 +1,111 @@
+ *
+ * Converts characters to lowercase, adds spaces around punctuation and multi-byte
+ * characters, strips accents from letters in the text and standardizes whitespace
+ * characters to all be the code point for the " " literal.
+ */
+class data_normalizer {
+ public:
+  /**
+   * @brief Transfer to the GPU the metadata needed to normalize characters.
+   *
+   * @param max_num_strings Maximum number of input strings for instantiating the normalizer.
+   *        Used to allocate temporary working memory on device.
+   *        If the input contains a larger number of strings, behavior is undefined.
+   * @param max_num_chars Maximum number of characters for instantiating the normalizer.
+   *        Used to allocate temporary working memory on device.
+   *        If the input contains a larger number of characters, behavior is undefined.
+   * @param do_lower_case If true, the normalizer will convert uppercase characters in the
+   *        input stream to lower case and strip accents from those characters.
+   *        If false, accented and uppercase characters are not transformed.
+   * @param stream CUDA stream used for device memory operations and kernel launches.
+   */
+  data_normalizer(uint32_t max_num_strings,
+                  uint32_t max_num_chars,
+                  cudaStream_t stream,
+                  bool do_lower_case = true);
+
+  /**
+   * @brief Normalize a vector of strings.
+   *
+   * If `do_lower_case` is true, this function will convert each character to lowercase
+   * and strip accents from the characters. If false it will do all other conversions
+   * in the class description except lower-casing and punctuation stripping.
+   *
+   * The result of this function returns two pointers to GPU data along with their lengths.
+   * The first pointer is to a contiguous array of unicode code points corresponding to the
+   * characters in the text after running normalization. The second pointer is to the
+   * offsets of the strings in the code point array. That is, string `i` starts at
+   * `result.second.gpu_ptr[i]`.
+   * This array will always be of length `num_strings + 1` since we need one entry
+   * for each input and a last entry which has the total number of bytes.
+   *
+   * @param d_strings A vector of strings which MUST be encoded in the UTF-8 format.
+   * @param d_offsets A vector of byte offsets to the beginning of individual strings in
+   *        the `d_strings` parameter.
+   * @param num_strings The number of strings identified in `d_strings`.
+   * @return Two pointers to GPU data along with their lengths. The first is a pointer
+   *         to the code points array and the second is a pointer to the offsets
+   *         used to locate the code points for each string.
+   * @param stream CUDA stream used for device memory operations and kernel launches.
+   */
+  std::pair<ptr_length_pair, ptr_length_pair> normalize(const char* d_strings,
```

Comment:
These pairs point to internal memory segments created and managed by the normalizer. It may be possible to allocate the memory using a uvector or buffer and return these instead but I feel that is a bigger rewrite than just porting the code as per this PR. I feel more complex improvements should go in a separate PR.